File: Jiho/week6/day26/미로 탈출 명령어.js
function solution(n, m, x, y, r, c, k) {
  let answer = "";
  // 이동방향을 사전순으로
  const direction = [
    [1, 0, "d"],
    [0, -1, "l"],
    [0, 1, "r"],
    [-1, 0, "u"],
  ];

  const distance = Math.abs(x - r) + Math.abs(y - c);

  // 도달 불가
  if (distance > k || (k - distance) % 2 !== 0) return "impossible";

  function dfs(x, y, path, depth) {
    if (depth + Math.abs(x - r) + Math.abs(y - c) > k) return;

    if (depth === k) {
      if (x === r && y === c) answer = path;
      return;
    }

    for (const [dx, dy, move] of direction) {
      const nx = x + dx,
        ny = y + dy;
      if (nx > 0 && nx <= n && ny > 0 && ny <= m) {
        dfs(nx, ny, path + move, depth + 1);
        if (answer) return;
      }
    }
  }

  dfs(x, y, "", 0);

  return answer;
}

File: Donggyo/week6/day26/미로탈출명령어.java
import java.util.*;

class Solution {  
    static int N,M,X,Y,R,C,K;
    static int[] dr = {1,0,0,-1}; // dlru
    static int[] dc = {0,-1,1,0};
    static String answer = "z";
    static String[] directions = {"d", "l","r","u"};
    
    public String solution(int n, int m, int x, int y, int r, int c, int k) {
        this.N=n;
        this.M=m;
        this.X=x;
        this.Y=y;
        this.R=r;
        this.C=c;
        this.K=k;
        
        if (getDistance(y,x) > K || (K - getDistance(y,x)) % 2 == 1){
            return "impossible";
        }
        
        dfs(X,Y,0,"");
        return answer.equals("z")? "impossible": answer;
    }
    
    static void dfs(int r, int c, int count, String path){
        if(count>K) return;
        if(path.compareTo(answer)>0) return;
        if(count+getDistance(r,c)>K) return;
        if(count == K && r==R && c==C ){
            answer = path;
            return;
        }
        
        for(int d=0;d<4;d++){
            int idr = r+dr[d];
            int idc = c+dc[d];
            
            if(idr<1||idc<1||idr>=N+1||idc>=M+1) continue;
            String newPath = path+directions[d];
            dfs(idr, idc, count+1, newPath);
        }
    }
    
    static int getDistance(int r, int c){
        return Math.abs(R-r)+Math.abs(C-c);
    }
}